libraries/hotstuff/include/eosio/hotstuff/qc_chain.hpp

```diff
@@ -0,0 +1,196 @@
+      void process_proposal(const hs_proposal_message& msg); //handles proposal
+      void process_vote(const hs_vote_message& msg); //handles vote
+      void process_new_view(const hs_new_view_message& msg); //handles new view
+      void process_new_block(const hs_new_block_message& msg); //handles new block
+
+      hs_vote_message sign_proposal(const hs_proposal_message& proposal, name finalizer); //sign proposal
+
+      bool extends(const fc::sha256& descendant, const fc::sha256& ancestor); //verify that a proposal descends from another
+
+      void on_beat(); //handler for pacemaker beat()
+
+      void update_high_qc(const eosio::chain::quorum_certificate& high_qc); //check if update to our high qc is required
+
+      void leader_rotation_check(); //check if leader rotation is required
+
+      bool is_node_safe(const hs_proposal_message& proposal); //verify if a proposal should be signed
+
+      std::vector<hs_proposal_message> get_qc_chain(const fc::sha256& proposal_id); //get 3-phase proposal justification
+
+      void send_hs_proposal_msg(const hs_proposal_message& msg); //send vote msg
+      void send_hs_vote_msg(const hs_vote_message& msg); //send proposal msg
+      void send_hs_new_view_msg(const hs_new_view_message& msg); //send new view msg
+      void send_hs_new_block_msg(const hs_new_block_message& msg); //send new block msg
+
+      void on_hs_vote_msg(const hs_vote_message& msg); //vote msg event handler
+      void on_hs_proposal_msg(const hs_proposal_message& msg); //proposal msg event handler
+      void on_hs_new_view_msg(const hs_new_view_message& msg); //new view msg event handler
+      void on_hs_new_block_msg(const hs_new_block_message& msg); //new block msg event handler
+
+      void update(const hs_proposal_message& proposal); //update internal state
+      void commit(const hs_proposal_message& proposal); //commit proposal (finality)
+
+      void gc_proposals(uint64_t cutoff); //garbage collection of old proposals
+
+private:
+
+      // This mutex synchronizes all writes to the data members of this qc_chain against
+      //   get_state() calls (which ultimately come from e.g. the HTTP plugin).
+      // This could result in a HTTP query that gets the state of the core while it is
+      //   in the middle of processing a given request, since this is not serializing
+      //   against high-level message or request processing borders.
+      // If that behavior is not desired, we can instead synchronize this against a
+      //   consistent past snapshot of the qc_chain's state for e.g. the HTTP plugin,
+      //   which would be updated at the end of processing every request to the core
+      //   that does alter the qc_chain (hotstuff protocol state).
+      // And if the chain_pacemaker::_hotstuff_global_mutex locking strategy is ever
+      //   changed, then this probably needs to be reviewed as well.
+      //
+      mutable std::mutex     _state_mutex;
+
+#ifdef QC_CHAIN_SIMPLE_PROPOSAL_STORE
```

Comment:
Have we decided on a proposal store? Right now `QC_CHAIN_SIMPLE_PROPOSAL_STORE` is not defined.

Reply:
The `QC_CHAIN_SIMPLE_PROPOSAL_STORE` and associated code will be removed. Maybe can be done during #1541 or #1520